pkg/kubelet/dockershim/docker_container.go

```diff
@@ -167,20 +167,25 @@ func (ds *dockerService) CreateContainer(_ context.Context, r *runtimeapi.Create
 	if err != nil {
 		return nil, err
 	}
-	defer func() {
-		for _, err := range ds.performPlatformSpecificContainerCreationCleanup(cleanupInfo) {
-			klog.Warningf("error when cleaning up after container %v's creation: %v", containerName, err)
-		}
-	}()
 
 	createResp, createErr := ds.client.CreateContainer(createConfig)
 	if createErr != nil {
 		createResp, createErr = recoverFromCreationConflictIfNeeded(ds.client, createConfig, createErr)
 	}
 
 	if createResp != nil {
-		return &runtimeapi.CreateContainerResponse{ContainerId: createResp.ID}, nil
+		containerID := createResp.ID
+
+		if cleanupInfo != nil {
+			// save it for when the container starts or gets removed
+			ds.containerCleanupInfos[containerID] = cleanupInfo
+		}
+		return &runtimeapi.CreateContainerResponse{ContainerId: containerID}, nil
 	}
+
+	// the creation failed, let's clean up right away
+	ds.performPlatformSpecificContainerCleanupAndLogErrors(containerName, cleanupInfo)
```

Comment:
previously this would have always been called in the defer, but now its not.  this seems awkwardly structured and/or confusing.  i need to see what this actually was cleaning up.

Reply:
actually, nevermind, i think this makes sense.

Reply:
@derekwaynecarr : better suggestions welcome :)

This is meant to handle any clean up that might be needed after a container creation - which as of now only is registry keys used to pass down registry keys down to Docker (see previous PR).

It used to be in a defer here, and so performed right after the container's creation; turned out that wasn't good enough, as Docker needs these reg keys to be present when it starts the container instead; so this clean up needs to happen after that.

Reply:
@wk8 could you make it clear in the comment that the key is needed for starting the container?

Reply:
@yujuhong : tried to make that clear at https://github.com/kubernetes/kubernetes/pull/74737/files#diff-c7dd39479fd733354254e70845075db5R180 ; please let me know if you think that's okay.